test/server/config_validation/xds_fuzz.cc

```diff
@@ -0,0 +1,250 @@
+ */
+absl::optional<std::string> XdsFuzzTest::removeListener(uint32_t listener_num) {
+  std::string match = fmt::format("{}{}", "listener_", listener_num % NUM_LISTENERS);
+
+  for (auto it = listeners_.begin(); it != listeners_.end(); ++it) {
+    if (it->name() == match) {
+      std::string name = it->name();
+      listeners_.erase(it);
+      return name;
+    }
+  }
+  return {};
+}
+
+/**
+ * remove a route from the list of routes if it exists
+ * @param the route number to be removed
+ * @return the route as an optional so that it can be used in a delta request
+ */
+absl::optional<std::string> XdsFuzzTest::removeRoute(uint32_t route_num) {
+  std::string match = fmt::format("{}{}", "route_config_", route_num % NUM_ROUTES);
+  for (auto it = routes_.begin(); it != routes_.end(); ++it) {
+    if (it->name() == match) {
+      std::string name = it->name();
+      routes_.erase(it);
+      return name;
+    }
+  }
+  return {};
+}
+
+/**
+ * run the sequence of actions defined in the fuzzed protobuf
+ */
+void XdsFuzzTest::replay() {
+  initialize();
+
+  // set up cluster
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}, {}, {}, true));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                             {buildCluster("cluster_0")},
+                                                             {buildCluster("cluster_0")}, {}, "1");
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "",
+                                      {"cluster_0"}, {"cluster_0"}, {}));
+  sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+      Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+      {buildClusterLoadAssignment("cluster_0")}, {}, "1");
+
+  version_++;
+
+  for (const auto& action : actions_) {
+    switch (action.action_selector_case()) {
+    case test::server::config_validation::Action::kAddListener: {
+      removeListener(action.add_listener().listener_num());
+      auto listener =
+          buildListener(action.add_listener().listener_num(), action.add_listener().route_num());
+      listeners_.push_back(listener);
+
+      updateListener(listeners_, {listener}, {});
+      // TODO(samflattery): compareDiscoveryResponse to check ACK/NACK?
+      break;
+    }
+    case test::server::config_validation::Action::kRemoveListener: {
+      auto removed = removeListener(action.add_listener().listener_num());
+
+      if (removed) {
+        updateListener(listeners_, {}, {*removed});
+      } else {
+        updateListener(listeners_, {}, {});
+      }
+
+      break;
+    }
+    case test::server::config_validation::Action::kAddRoute: {
+      removeRoute(action.add_route().route_num());
+      auto route = buildRouteConfig(action.add_route().route_num());
+      routes_.push_back(route);
+      updateRoute(routes_, {route}, {});
```

Comment:
Just curious. Here if the route already exists, it looks like push an update with the same route config. Is that correct?